### PR TITLE
chore: upgrade to V21 and fix Lit 2 usage and TS event names

### PR DESCRIPTION
**Pull request by denis-anisimov**
fixes vaadin/flow#10872

**Update**
Here's a summary of what this PR does.
- Upgrades Vaadin version to V21 and does updates required by breaking changes (in Lit, flow and our components)
  - Fix/update Lit related import paths in TS files and example snippets (additionally make these imports a bit more consistent in code style and ordering since we're touching most of the files anyway)
  - Replace `@internalProperty` decorator with `@state` (this is a rename in Lit API)
  - Fix TS compilation by updating Java classes related to breaking nullability changes in vaadin/flow#8849
  - Use new Vaadin Component TS events as required due to this breaking change vaadin/web-components#265
  - Fix `applyTheme()` usage due to breaking change in Lit 2 (calling it in the constructor does not work anymore as it is now too early. `createRenderRoot()` is now called later by Lit so `this.shadowRoot` is not available in the constructor)
- Adds a section into `upgrading.asciidoc` about Lit 2 related changes

### CODE REVIEW

File: articles/guide/upgrading.asciidoc [hunk 1]
```diff
@@ -42,10 +42,74 @@ include::{root}/pom.xml[tag=spring-version, indent=0]
 
 . Run `mvn clean install`
 
+== Changes in Vaadin 21
+
+These instructions apply when upgrading from a version before Vaadin 21.
+
+=== Lit Templates are using Lit 2
```
Comment (haijian-vaadin) [inline articles/guide/upgrading.asciidoc:49]: Change to title case? Lit Templates Are Using Lit 2, as reported by Vale?

**Review by haijian-vaadin — CHANGES_REQUESTED** [review]

File: articles/guide/upgrading.asciidoc [hunk 2]
```diff
@@ -42,10 +42,74 @@ include::{root}/pom.xml[tag=spring-version, indent=0]
...
+
+Previously, Lit templates were bases on LitElement 2.x and lit-html 1.x.
+The new Lit 2.0 is used for Lit templates in the latest Vaadin version.
```
Comment (magi42) [inline articles/guide/upgrading.asciidoc:52]:
```suggestion
New Lit 2.0 is used for Lit templates in the latest Vaadin version.
```

File: articles/guide/upgrading.asciidoc [hunk 3]
```diff
@@ -42,10 +42,74 @@ include::{root}/pom.xml[tag=spring-version, indent=0]
...
+----
+import { repeat } from 'lit-html/directives/repeat.js';
+----
+
+to
+
+[source,typescript]
+----
+import { repeat } from 'lit/directives/repeat.js';
+----
```
Comment (magi42) [inline articles/guide/upgrading.asciidoc:102]:
```suggestion
Most of the necessary changes are with imports.

You need to change imports of `html` and [classname]#LitElement# from:

[source,typescript]
----
import { html, LitElement } from 'lit-element';
----

to:

[source,typescript]
----
import { html, LitElement } from 'lit';
----

Change `customElement` and `property` decorator imports from:

[source,typescript]
----
import { customElement, property } from 'lit-element';
----

to

[source,typescript]
----
import { customElement, property } from 'lit/decorators.js';
----

If you are using `repeat` directive imports, change them from:

[source,typescript]
----
import { repeat } from 'lit-html/directives/repeat.js';
----

to

[source,typescript]
----
import { repeat } from 'lit/directives/repeat.js';
----
```

File: articles/guide/upgrading.asciidoc [hunk 4]
```diff
@@ -42,10 +42,74 @@ include::{root}/pom.xml[tag=spring-version, indent=0]
...
+
+Some Lit APIs have been renamed.
+The most noticeable change is that the `@internalProperty` decorator has been renamed to `@state`.
+
+
 == Changes in Vaadin 20
 
 These instructions apply when upgrading from a version before Vaadin 20.
```
Comment (magi42) [inline articles/guide/upgrading.asciidoc:110]:
```suggestion
The following instructions apply when upgrading from a version before Vaadin 20.
```

**Review by magi42 — CHANGES_REQUESTED** [review]
Reviewed `upgrading.asciidoc`. A bit delicate to do without preview, so need to check that carefully.

File: articles/guide/upgrading.asciidoc (hunk 3 above)
Comment (Haprog) [reply articles/guide/upgrading.asciidoc:101]: These imports are just examples of different type of imports and how they should be changed. This is not an exhaustive list so mentioning only "Change `customElement` and `property` decorator imports" or "If you are using `repeat` directive imports" explicitly is not good.

I'll use your suggestion as a basis but make some changes to the text content to make it more clear.

Comment (magi42) [reply articles/guide/upgrading.asciidoc:101]: OK, the main point was that the text was formatted like it was a really long sentence with code examples in between.

Comment (Haprog) [reply articles/guide/upgrading.asciidoc]: Updated.

**Comment by Haprog** [comment]
@magi42 Thanks for the AsciiDoc review! I applied most of your suggestions directly and did some minor updates to the text content about updating the Lit imports.

**Review by haijian-vaadin — APPROVED** [review]

File: articles/guide/upgrading.asciidoc [hunk 5]
```diff
@@ -44,57 +44,55 @@ include::{root}/pom.xml[tag=spring-version, indent=0]
...
 Some changes are required to be able to use your existing template with Lit 2.0.
-Refer to the https://lit.dev/docs/releases/upgrade/#update-packages-and-import-paths[Upgrade guide]
-to find all necessary changes.
-In most cases you should be aware of changes in the imports:
+Please refer to the https://lit.dev/docs/releases/upgrade/#update-packages-and-import-paths[Lit Upgrade Guide] to find all the necessary changes.
+Most of the necessary changes are with imports.
 
-from
+Use the `lit` package.
+For example change imports of `html` and [classname]#LitElement# from:
```
Comment (magi42) [inline articles/guide/upgrading.asciidoc:58]:
```suggestion
For example, change imports of `html` and [classname]#LitElement# from:
```

File: articles/guide/upgrading.asciidoc [hunk 6]
```diff
@@ -44,57 +44,55 @@ include::{root}/pom.xml[tag=spring-version, indent=0]
...
 ----
 import { html, LitElement } from 'lit';
 ----
 
-decorator imports
-
-from
+Update decorator imports.
+Decorators have been moved to a separate module.
+For example change `customElement` and `property` decorator imports from:
```
Comment (magi42) [inline articles/guide/upgrading.asciidoc:74]:
```suggestion
For example, change `customElement` and `property` decorator imports from:
```

**Review by magi42 — CHANGES_REQUESTED** [review]
Just couple of commas missing there. Otherwise looks great.

**Comment by Haprog** [comment]
@magi42 I committed your latest suggestions.

**Review by magi42 — APPROVED** [review]

**Review by Haprog — APPROVED** [review]

**Pull request merged**